### PR TITLE
[SYCL][Matrix] Add more operators for wi_element

### PR DESCRIPTION
**Pull request by yubingex007-a11y**
1. We add binary operators: +, -, *, /=, and binary assignment operators
+=, -=, /=
2. Add support for logical operators: >, >=, <, <=, ==, !=
3. Add support for explicit operator bool()

### CODE REVIEW

**Review by dkhaldi — APPROVED** [review]
LGTM but just add a description that this is a continuation to the merged patch that adds more operator overloading to enable element wise operations for joint_matrix

File: sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp [hunk 1]
```diff
@@ -259,6 +338,99 @@ class wi_element {
...
+#ifdef __SYCL_DEVICE_ONLY__
+    M.spvm = __spirv_VectorInsertDynamic(
+        M.spvm, __spirv_VectorExtractDynamic(M.spvm, idx) / rhs, idx);
+    return *this;
+#else
+    (void)rhs;
+    throw runtime_error("joint matrix is not supported on host device.",
+                        PI_INVALID_DEVICE);
+#endif // __SYCL_DEVICE_ONLY__
+  }
```
Comment (steffenlarsen) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp:362]: Implementation looks good, but if you feel like avoiding replicated code I believe the following should work as well. It applies to all the other similar operators as well. Just a suggestion. 😄 
```suggestion
  friend wi_element &operator/=(wi_element<T, NumRows, NumCols, Layout, Group> &lhs, const T &rhs) {
#ifdef __SYCL_DEVICE_ONLY__
    lhs = lhs / rhs;
    return lhs;
#else
    (void)lhs;
    (void)rhs;
    throw runtime_error("joint matrix is not supported on host device.",
                        PI_INVALID_DEVICE);
#endif // __SYCL_DEVICE_ONLY__
  }
```
You could even just have the `__SYCL_DEVICE_ONLY__` case and rely on `/` to fail on host, though that might make the stack trace a little confusing. I am not sure it matters in this case however.

File: sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp [hunk 2]
```diff
@@ -248,6 +258,75 @@ class wi_element {
                         PI_INVALID_DEVICE);
 #endif // __SYCL_DEVICE_ONLY__
   }
+  wi_element &
```
Comment (steffenlarsen) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp:261]: For the sake of visual distinction I would prefer a space between each operator overload.

File: sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp [hunk 3]
```diff
@@ -248,6 +258,75 @@ class wi_element {
...
+                        PI_INVALID_DEVICE);
+#endif // __SYCL_DEVICE_ONLY__
+  }
+
+  friend T operator+(const wi_element<T, NumRows, NumCols, Layout, Group> &lhs,
+                     const T &rhs) {
+#ifdef __SYCL_DEVICE_ONLY__
+    return __spirv_VectorExtractDynamic(lhs.M.spvm, lhs.idx) + rhs;
+#else
+    (void)rhs;
```
Comment (steffenlarsen) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp:279]: Is there a reason not to ignore `lhs` here and in other overloads with it?

Comment (yubingex007-a11y) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp:279]: Nice catch, lhs should not be ignored

File: sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp (hunk 2 above)
Comment (yubingex007-a11y) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp:262]: Sure, I've added them

File: sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp (hunk 1 above)
Comment (yubingex007-a11y) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp:369]: I think it is a nice catch, but we hope merge it soon. so, Maybe later i will address it

**Review by dkhaldi — APPROVED** [review]

**Comment by yubingex007-a11y** [comment]
@intel/llvm-reviewers-runtime ping?

**Review by steffenlarsen — APPROVED** [review]
LGTM!

**Comment by yubingex007-a11y** [comment]
@intel/llvm-reviewers-runtime  ping?

**Pull request merged**